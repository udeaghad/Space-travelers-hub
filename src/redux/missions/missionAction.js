import { createAsyncThunk } from '@reduxjs/toolkit';

const baseUrl = 'https://api.spacexdata.com/v3/missions';

const GET_ALL_MISSION = 'GET_ALL_MISSION';

const JOIN_MISSION = 'JOIN_MISSION';

const getMissions = createAsyncThunk(
  GET_ALL_MISSION,
  async () => {
    const request = new Request(`${baseUrl}`);
    const response = await fetch(request);
    const result = await response.json();

    const missionDetails = [];
    result.forEach((mission) => {
      const { description, mission_name, mission_id } = mission;
      missionDetails.push({
        description, mission_name, mission_id, join_mission: false,
      });
    });

    return missionDetails;
  },
);

const joinMission = (payload) => ({
  type: JOIN_MISSION,
  payload,
});

export { getMissions, joinMission };
