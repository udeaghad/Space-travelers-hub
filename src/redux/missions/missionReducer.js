const mission = [];

export default (state = mission, action) => {
  switch (action.type) {
    case 'GET_ALL_MISSION/fulfilled':
      
      return action.payload;
    
    case 'JOIN_MISSION':
     state.map(item => {
      if(item.mission_id === action.payload){
        return {
          ...state,
          join_mission: !state.join_mission,
        };
      }
     })

     default:
      return state
  }
}

