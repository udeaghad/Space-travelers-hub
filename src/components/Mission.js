import {useDispatch, useSelector} from 'react-redux'
import React, { useEffect } from 'react';
import { getMissions, joinMission } from '../redux/missions/missionAction';



const Mission = () => {
const showMission = useSelector(state => state.Mission)

const dispatch = useDispatch()

useEffect(() => {
  if(!showMission.length){
    dispatch(getMissions())
}
}, [])


const handleClick = (e) => {
  e.preventDefault();
  let payload = e.target.id
  dispatch(joinMission(payload))
  }

return (
  <div > 
       
      {showMission.map(mission => (      
        
        <ul className='wrapper' key={mission.mission_id}>
         <li className='name-cell'>{mission.mission_name}</li>
         <li className='description-cell'>{mission.description}</li> 
          {!mission.join_mission?
          <div>
          <span className='badge'>NOT A MEMBER</span>
          <button type='button' id={mission.mission_id} onClick={(e) => handleClick(e)}>JOIN MISSION</button>
          </div>: 
          <div>
            <span className='badge'>ACTIVE MEMBER</span>
          <button type='button' id={mission.mission_id} onClick={(e) => handleClick(e)}>LEAVE MISSION</button>
          </div>
          }
        
        </ul>
      
       
        ))}
   
</div>
)

}

export default Mission