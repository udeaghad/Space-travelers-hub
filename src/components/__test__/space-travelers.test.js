import missionReducer from '../../redux/missions/missionReducer'
import {joinMission} from '../../redux/missions/missionAction'
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../redux/configureStore';
import Mission from '../Mission';


describe('To test Mission components', () => {
  test('To Missions component renders successfully!', () => {
    const tree = renderer
      .create(<Provider store={store}><Mission /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
})

test('To test join mission status', () => {
  const Mission = [
    {mission_id:1, mission_name: "Teslar", description: "Telstar 19V (Telstar 19 Vantage)...", join_mission: false},
    {mission_id:2, mission_name: "TThaicom", description: "Thaicom is the name of a series...", join_mission: false},
    {mission_id:3, mission_name: "JCSAT", description: "The JSAT constellation is a...", join_mission: false},
  ]

  expect(missionReducer(Mission, joinMission(2))).toEqual(
    [
      {mission_id:1, mission_name: "Teslar", description: "Telstar 19V (Telstar 19 Vantage)...", join_mission: false},
      {mission_id:2, mission_name: "TThaicom", description: "Thaicom is the name of a series...", join_mission: true},
      {mission_id:3, mission_name: "JCSAT", description: "The JSAT constellation is a...", join_mission: false},
    ]
  )
})

});


