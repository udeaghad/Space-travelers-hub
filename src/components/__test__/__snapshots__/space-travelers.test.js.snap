// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To test Mission components To Missions component renders successfully! 1`] = `<div />`;
